junit5-jupiter-extensions/src/main/java/com/example/cartesian/CartesianProductContext.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2015-2018 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package com.example.cartesian;
+
+import java.util.Collections;
+import java.util.List;
+import org.junit.jupiter.api.extension.Extension;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolver;
+import org.junit.jupiter.api.extension.TestTemplateInvocationContext;
+
+public class CartesianProductContext implements TestTemplateInvocationContext, ParameterResolver {
+
+	private final List<?> parameters;
+
+	CartesianProductContext(List<?> parameters) {
+		this.parameters = parameters;
+	}
+
+	@Override
+	public String getDisplayName(int invocationIndex) {
+		return invocationIndex + ": " + parameters.toString();
+	}
+
+	@Override
+	public List<Extension> getAdditionalExtensions() {
+		return Collections.singletonList(this);
```

Comment:
I don't think it's a good idea for an extension to register _itself_ again in the extension registry, since that could potentially lead to undesirable side effects.

This may work in this scenario, but I fear others might think it's a "best practice" since it's included in an example by core JUnit committers.

I'd recommend implementing the `ParameterResolver` API in its own class.